content/2021-06-11-make-your-life-easier.md

```diff
@@ -0,0 +1,156 @@
+anything containing [personal data](https://gdpr-info.eu/art-4-gdpr/) must be
+deletable. So just like with graph data structures there's a kind of hierarchy
+of mutability of data structures:
+
+- Truly immutable. The data is created and is never destroyed.
+- Can be atomicly created or destroyed but never modified.
+- Can have portions appended or deleted, but previous data is never modified.
+- Some portions of the data structure are immutable, others can be deleted or
+  mutated in place.
+- All parts of the data structure are mutable.
+
+Stay as far up that list as you can and you will avoid many implementation
+problems.
+
+# Choose simpler programming models
+
+There are lots of techniques to improve the performance of software. Use none
+of them until you are forced to. Concurrency is hard to get right.
+Event-driven systems can be difficult to reason about. Manual memory
+management is a notorious source of bugs. Functions without side-effects are
+easier to test than functions with them.
+
+Start with single-threaded, single process code in the highest-level language
+you can use and iterate from there. Use pure functions when you can, but don't
+be a zealot about it. The ["Functional Core, Imperative
+Shell"](https://www.destroyallsoftware.com/screencasts/catalog/functional-core-imperative-shell)
+pattern can be very useful.
+
+# Worry about all your public interfaces
+
+[Hyrum's law](https://www.hyrumslaw.com/) teaches us that anything your
+customer can perceive about your software is part of its public interface. If
+you change it, [someone will complain](https://xkcd.com/1172/).
+
+When you're designing new software, divide your problem space into features
+and behaviours that are visible and invisible to customers. Put the bulk of
+your effort into the visible parts, everything else is fixable later.
+
+If you don't want your customer to depend on something, make it impossible to
+depend upon. For example, if you have a function that returns a list of items
+and for some reason you don't want the ordering to be defined then you may
+need to deliberately shuffle the items to prevent people from depending on the
+order.
+
+# You probably can't avoid distributed systems
+
+Even if you have only one web server running a monolithic application, if you
+put a rich JavaScript web app in front of it you have a distributed system. If
+you only work on static, single binary, CLI tools that don't touch the network
+then you can avoid distributed systems but you might also be severely limiting
+your career.
+
+It's extremely hard to avoid distributed systems so you have to get good at
+working on them. Here are some things to remember:
+
+- There's no such thing as a zero-downtime atomic deploy. You can choose
+  downtime if you want, but choosing a safe, non-atomic deployment pattern is
+  much easier.
+- Your system's topology is a graph. The guidance for graph data structures
```

Comment:
Ohhhhh, the aforementioned graph!

Reply:
How can you control dependencies in a distributed system? I mean, as an individual engineer, how much control do you have over other folks putting your service into a cycle?

Reply:
That's more of a social problem than a technical problem. As an individual engineer you only have control in the context of being able to persuade your colleagues to do the right thing.